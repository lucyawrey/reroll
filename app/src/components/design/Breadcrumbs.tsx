import React from "react";
import { Breadcrumb as BSBreadcrumb } from "react-bootstrap";
import Link from "next/link";
import { useRouter } from "next/router";

interface BreadcrumbProps {
  skipLevels?: number;
  titles?: (string | undefined)[];
}

/**
 * A simple breadcrumb component that builds out the links from the URL.
 * @param props.skipLevels A number indicating how many levels to skip. Default 0
 * @param props.titles An array of strings for the title of each level
 */
function Breadcrumbs(props: BreadcrumbProps): JSX.Element {
  const router = useRouter();
  const skipLevels: number = props.skipLevels || 0;
  const titles: (string | undefined)[] = props.titles || [];

  /**
   * Builds out the breadcrumb links
   */
  function buildCrumbs() {
    // TODO - BUG where trailing / causes extra slash to appear in breadcrumbs
    const crumbs = router.asPath.split("/");
    const breadcrumbs: JSX.Element[] = [];
    let uri = "/";

    for (let i = 0; i < crumbs.length; i++) {
      const titleIndex: number = i - skipLevels;
      uri += crumbs[i];

      if (titleIndex < 0) {
        continue;
      }

      let title: string = crumbs[i];
      if (titles.length >= titleIndex) {
        title = titles[titleIndex] || "";
      }

      breadcrumbs.push(
        <Link key={"crumb_" + i} href={uri} passHref>
          <BSBreadcrumb.Item>{title}</BSBreadcrumb.Item>
        </Link>,
      );

      if (i < crumbs.length - 1) {
        uri += "/";
      }
    }

    return breadcrumbs;
  }

  return (
    <BSBreadcrumb>
      { buildCrumbs() }
    </BSBreadcrumb>
  );
}

export default Breadcrumbs;
