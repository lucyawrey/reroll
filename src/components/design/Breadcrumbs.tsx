import React from "react";
import { Breadcrumb as BSBreadcrumb } from "react-bootstrap";
import Link from "next/link";
import { useRouter } from "next/router";
import { def } from "../../utilities/tools";

interface BreadcrumbProps {
  skipLevels?: number; // the endpoint to start from
  titles?: string[];
}

/**
 * A simple to use breadcrumb component
 * @param props see IBreadcrumb
 */
function Breadcrumbs(props: BreadcrumbProps) {
  const router = useRouter();
  const skipLevels: number = def<number>(props.skipLevels, 0);
  const titles: string[] = def<string[]>(props.titles, []);

  /**
   * Builds out the breadcrumb links
   */
  function buildCrumbs() {
    // TODO - BUG where trailing / causes extra slash to appear in breadcrumbs
    const crumbs = router.asPath.split("/");
    const breadcrumbs: JSX.Element[] = [];
    let uri = "/";

    for (let i: number = 0; i < crumbs.length; i++) {
      const titleIndex: number = i - skipLevels;
      uri += crumbs[i];

      if (titleIndex < 0) {
        continue;
      }

      let title: string = crumbs[i];
      if (titles.length >= titleIndex) {
        title = titles[titleIndex];
      }

      breadcrumbs.push(
        <Link key={"crumb_" + i} href={uri} passHref>
          <BSBreadcrumb.Item>{title}</BSBreadcrumb.Item>
        </Link>,
      );

      if (i < crumbs.length - 1) {
        uri += "/";
      }
    }

    return breadcrumbs;
  }

  return (
    <BSBreadcrumb>
      { buildCrumbs() }
    </BSBreadcrumb>
  );
}

export default Breadcrumbs;
