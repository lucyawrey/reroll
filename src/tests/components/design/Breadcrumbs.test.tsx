import React from "react";
import renderer from "react-test-renderer";
import Breadcrumbs from "components/design/Breadcrumbs";
import { WithTestRouter } from "utilities/tests/WithTestRouter";

const push = jest.fn();

describe('Breadcrumb rendering', () => {
  it('renders correctly with an admin path', () => {
    const router = {
      push,
      pathname: "/admin/game-systems/[gameSystemAlias]/modules/[moduleAlias]",
      asPath: "/admin/game-systems/dnd-5e/modules/srd",
    };
    const tree = renderer.create(
      <WithTestRouter router={router}>
        <Breadcrumbs
          skipLevels={1}
          titles={["Game Systems", "Dungeons & Dragons, 5e", "Modules", "SRD"]}
        />
      </WithTestRouter>
    ).toJSON();


    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with no arguments', () => {
    const router = {
      push,
      pathname: "/admin/game-systems/[gameSystemAlias]/modules/[moduleAlias]",
      asPath: "/admin/game-systems/dnd-5e/modules/srd",
    };
    const tree = renderer.create(
      <WithTestRouter router={router}>
        <Breadcrumbs />
      </WithTestRouter>
    ).toJSON();


    expect(tree).toMatchSnapshot();
  });
});

