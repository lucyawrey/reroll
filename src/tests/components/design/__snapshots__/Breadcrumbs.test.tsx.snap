// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breadcrumb rendering renders correctly with an admin path 1`] = `
<nav
  aria-label="breadcrumb"
>
  <ol
    className="breadcrumb"
  >
    <li
      className="breadcrumb-item"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      <a
        href="/admin"
        onClick={[Function]}
        onKeyDown={[Function]}
      >
        Game Systems
      </a>
    </li>
    <li
      className="breadcrumb-item"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      <a
        href="/admin/game-systems"
        onClick={[Function]}
        onKeyDown={[Function]}
      >
        Dungeons & Dragons, 5e
      </a>
    </li>
    <li
      className="breadcrumb-item"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      <a
        href="/admin/game-systems/dnd-5e"
        onClick={[Function]}
        onKeyDown={[Function]}
      >
        Modules
      </a>
    </li>
    <li
      className="breadcrumb-item"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      <a
        href="/admin/game-systems/dnd-5e/modules"
        onClick={[Function]}
        onKeyDown={[Function]}
      >
        SRD
      </a>
    </li>
    <li
      className="breadcrumb-item"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      <a
        href="/admin/game-systems/dnd-5e/modules/srd"
        onClick={[Function]}
        onKeyDown={[Function]}
      >
        
      </a>
    </li>
  </ol>
</nav>
`;

exports[`Breadcrumb rendering renders correctly with no arguments 1`] = `
<nav
  aria-label="breadcrumb"
>
  <ol
    className="breadcrumb"
  >
    <li
      className="breadcrumb-item"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      <a
        href="/"
        onClick={[Function]}
        onKeyDown={[Function]}
      >
        
      </a>
    </li>
    <li
      className="breadcrumb-item"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      <a
        href="//admin"
        onClick={[Function]}
        onKeyDown={[Function]}
      >
        admin
      </a>
    </li>
    <li
      className="breadcrumb-item"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      <a
        href="//admin/game-systems"
        onClick={[Function]}
        onKeyDown={[Function]}
      >
        game-systems
      </a>
    </li>
    <li
      className="breadcrumb-item"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      <a
        href="//admin/game-systems/dnd-5e"
        onClick={[Function]}
        onKeyDown={[Function]}
      >
        dnd-5e
      </a>
    </li>
    <li
      className="breadcrumb-item"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      <a
        href="//admin/game-systems/dnd-5e/modules"
        onClick={[Function]}
        onKeyDown={[Function]}
      >
        modules
      </a>
    </li>
    <li
      className="breadcrumb-item"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      <a
        href="//admin/game-systems/dnd-5e/modules/srd"
        onClick={[Function]}
        onKeyDown={[Function]}
      >
        srd
      </a>
    </li>
  </ol>
</nav>
`;
